### DISCUSSION TITLE
Serial control discussion and gallery

### DISCUSSION THREAD

**Discussion by DJPoulter** [gcormier/megadesk#87] [General]
Reading through #12 I couldn't seem to find the best solution to this so asking it here. But what are peoples recommendations on a controller for serial control? I was thinking of just using a D1 mini. 

And then to follow it up would the best solution be to run connect the 3 pins from the programming header and run the cable out to the controller? Or has someone worked out a way to get power and fit a controller in the box itself?

**Comment by gcormier** [comment]
converting to Discussion

**Reply by DJPoulter** [reply]
Ah sorry forgot discussions were a thing

**Comment by gcormier** [comment]
I would definitely not recommend tapping the megadesk for power. The thermal implications probably aren't great as linear regulators are used, which produce heat to drop the 30V down to 12V and 5V. This is OK as our consumption is very low - about 15mA.

Adding something that consumes 10 the power will be too much heat inside the enclosure.

The new PCB version (just some layout changes, nothing new) actually has silkscreen now that says not to use the power for any other peripherals since more and more folks are interested in the serial stuff.


I would suggest you could tap RX/TX/GND/24V from the megadesk, and then wherever it is going, have your own component to go from 24V to whatever voltage is needed by your accessory board.

**Comment by DJPoulter** [comment]
Okay perfect, my Megadesk should be arriving soon. I am keen to get tinkering with it.

**Comment by Will1604** [comment]
Unsure if adding a Bluetooth LE control option is off topic for this discussion, but going for it anyway.

The [Bluefruit LE UART](https://learn.adafruit.com/introducing-the-adafruit-bluefruit-le-uart-friend) looks possible. The question of flow control may or may not be an issue depending on circumstances.

The [current measurements](https://learn.adafruit.com/introducing-the-adafruit-bluefruit-le-uart-friend/current-measurements)  show average current: 1.86mA, peak current: 15.2mA, More like 2x power than the 10x for Wifi options. How does providing that via the programming header seem?

For the "tap RX/TX/GND/24V from the megadesk" route. Any suggestions on options for the 24V. Simplest might be a short breakout cable with 2.54mm headers between megadesk and the power cable. Details on using a DC Buck in issue #58 would complete the picture.

**Reply by gcormier** [reply]
2mA average consumption should be ok.

For tapping, yeah I think a 2.54mm cable is easiest. Either make your own completely, or something *like* these https://www.amazon.ca/Spektrum-Futaba-Harness-Splitter-Airplane/dp/B07PHDYZXN (I don't know if these will work as the RC ones I have have 2 rounded edges on the black housing)

**Comment by gcormier** [comment]
https://www.ezsbc.com/product/wifi01-sw/

Not affiliated but I found this board that has a High Voltage input and a buck converter built in, so you can power this from the 24V (30V) provided by the desk.

**Comment by fellwell5** [comment]
Hello @ all!

I managed to get my Megadesk running with an ESP8266 and connected it to my Homeassisant.
I wrote it up on a medium post so that other people can follow my steps:
https://medium.com/@matthiasschaffer/make-your-ikea-standing-desk-smart-62b8307da8b6

Also here is direct link to a case a friend of mine made for me, which fits a StepDown Converter or either an ESP8266 or an ESP32:
https://www.thingiverse.com/thing:5438136
![image](https://user-images.githubusercontent.com/4845418/180211904-02b2e0ae-4429-45bc-a2aa-f0fba7670a89.png)


Have a great day!

**Reply by gcormier** [reply]
Very awesome! Glad the servo splitter worked!

**Reply by AnAnalogGuy** [reply]
Nice work @fellwell5 .

Regarding:

![grafik](https://user-images.githubusercontent.com/70239317/221823362-86af34ee-344b-4d41-820d-7258add97654.png)
https://medium.com/@matthiasschaffer/make-your-ikea-standing-desk-smart-62b8307da8b6

Do I get it right and that's not needed any more with new orders of Megadesk as it now comes with the serial option enabled (if so, it might be helpful to update the medium article, so ppl know they don't need to flash the AT and therefor also don't need to by the related parts)? Or are there any other/additonal reasons to flash it?

**Reply by gcormier** [reply]
Correct, the serial firmware is now flashed by default since December 2022.

**Comment by Krasnov777** [comment]
I also received my Megadesk v.8 with serial enabled feature and decided to use Wemos D1 mini as a microcontroller. I used this [instruction](https://github.com/gcormier/megadesk/blob/master/esphome.md) for connecting and uploading configuration to the D1. 

Because this configuration just runs web server on the microcontroller, I realised that it's good idea to control the desk with HTTP POST method, and as a platform I used Apple Shortcuts and Telegram bot. Here few screenshots how I configured Apple Shortcuts:

![image](https://user-images.githubusercontent.com/62969603/217831069-aed963e2-dc0d-4ea3-8b53-cd83d2f67f1b.png)

And in the recent version of MacOS it's possible to add this shortcuts directly to your dock, so it's really great way to control your desk:
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/62969603/217831462-4a07ea62-d2f8-42aa-b96a-638a3c473f1a.png">

**Reply by gcormier** [reply]
Really cool, thanks for sharing!

**Reply by AnAnalogGuy** [reply]
@Krasnov777 
Perfect, that seems to be exactly what i was looking for.

Did you implement the control of megadesk's serial port and the mapping to http commands yourself? Arduino sketch using ESP8266WiFi.h? If so, would you be willing to share your source with the community?

**Reply by Krasnov777** [reply]
No, I'm just used [this configuration](https://github.com/gcormier/megadesk/blob/master/esphome.md), which runs web server on your ESPHome controller. 
![image](https://user-images.githubusercontent.com/62969603/221840317-4fbe25a2-8f29-4800-8cd6-948afaeca8ba.png)

On this web page you can press the buttons which mapped with the megadesk memory positions. So I just figured which POST URL used when you press those buttons and send it directly from Apple Shortcuts.

For instance, I saved in Megadesk memory positions for button 2 and 3 (for standing and sitting):

And here is how POST URL's looks like

```
1. http://YOURESPHOMEIP/button/desk_position_2/press
2. http://YOURESPHOMEIP/button/desk_position_3/press
```

**Reply by AnAnalogGuy** [reply]
Thanks a lot for clarifcation @Krasnov777.

**Reply by osb99** [reply]
Please can you send me this shortcut?

**Reply by Krasnov777** [reply]
@osb99  Please use this shortcut as a reference. You will need to change the IP and button_id.

https://www.icloud.com/shortcuts/10e0c8dc35784018bdd026f52f0f9eea

**Reply by osb99** [reply]
@Krasnov777 that's great, thanks!

**Comment by AnAnalogGuy** [comment]
Quick summary guide for those who want to run the ESP controlled megadesk (on Windows):

1. Get a megadesk from Greg's Tindie store: https://www.tindie.com/products/gcormier/megadesk/
2. Follow Greg's installation instructions: https://www.youtube.com/watch?v=jCPlM2KYwDQ. 
3. Get the ESP parts, mainly "D1 Mini ESP8266 WLAN Microcontroller" and a "LM2596S DC" (stepdown) converter. Any other ESP controller would work too, but this one has an USB interface, so you don't need a programmer, just a Micro-USB cable, that's it. 
4. An enclosure and some wires are needed too, check out Matthias' guide: https://medium.com/@matthiasschaffer/make-your-ikea-standing-desk-smart-62b8307da8b6
5. Get [megadesk.h](https://github.com/gcormier/megadesk/blob/master/esphome/megadesk.h) and [esphome.yaml](https://github.com/gcormier/megadesk/blob/master/esphome/megadesk.yaml) from the repository. 
6. In esphome.yaml replace the ssid and the password with the correct values from your WLAN router. Both values need to be enclosed with `""`. If you use `"` within your password, you have to escape it with `\"`. Value for the board parameter needs to be  "d1_mini" without quotes.
7. Download https://cdn.sparkfun.com/assets/learn_tutorials/5/9/7/Windows-CH340-Driver.zip and install it - it's an USB driver needed for Windows to communicate with the ESP board. 
8. Connect the ESP board to your PC via a Micro-USB cable. 
9. Get Python and install it: https://www.python.org/downloads/windows/. Add Python to the PATH variable of your Windows Computer. 
10. Open a windows cmd shell to install ESPHome in your Python (runtime) environment. Run
_pip3 install wheel
pip3 install esphome_
11. From the cmd compile and install the ESPHome package on your ESP microcontroller: 
_esphome run megadesk.yaml_
12. At the end of the installation (once it is complete), you can find the MAC address of the ESP microcontroller. Add it to your DHCP server and link it to an IP address (xxx.xxx.xxx.xxx). 
Alternative: You could also set a fixed IP address in the yaml configuration. 
13. Once done, restart your ESP microcontroller and point a browser to http://xxx.xxx.xxx.xxx to get access to the control interface

Notes: 
- I left out the wiring between the ESP controller and Megadesk, as Matthias has described it in detail on his medium article: https://medium.com/@matthiasschaffer/make-your-ikea-standing-desk-smart-62b8307da8b6
- I will add some more information on how to smoothly integrate with windows (clickable icons in the task bar just as @Krasnov777 did for Mac) and maybe with keyboard shortcuts. Have to figure out a solution first, will share. 
- I'm also thinking about a little Mobile App to control the desk, maybe also adding a timer function to remind users to change position after a while. Again, I have to figure out a solution first, will share, might take a while.

Keep in mind:
- ESPHome does not offer HTTPs support (reason: limited cpu power). This means, communication between the client app and the ESP board is unencrypted. 
- You could add a username/password protection to the webserver to add a certain level of protection, however, as basicauth is used, passwords are not encrypted and transported over non-secure http. 
- So bottom line: Consider your megadesk to be publicly acccessible, at least within your LAN. 

web_server:
  port: 80
  auth:
    username: admin
    password: !secret web_server_password

**Reply by gcormier** [reply]
Sure, I made #127 and will put some files up this week. Thanks for the step-by-step!

**Reply by bjornand1** [reply]
@AnAnalogGuy, I realize this recipe you posted is more than two years old, but I am looking for something similar when running on Home Assistant Green. I have no experience whatsoever with ESPHome, this is my very first attempt at getting into this world, and I am fairly lost...

I have installed ESPHome Builder in Home Assistant and have tried to install the Megadesk Companion as a new device, but I just don't know how to get the megadesk-companion.yaml flashed to the device. I have tried adding it in ESPHome Builder, but I see no option to upload that file...

**Reply by AnAnalogGuy** [reply]
Why don't you follow my guide above. It still should be working this way.

**Reply by bjornand1** [reply]
Thanks for your extremely quick reply! I would love to follow the guide, but I am on a Mac and running Home Assistant on a Home Assistant Green, so I figured I would do this from inside the ESPHome Builder interface?

**Reply by AnAnalogGuy** [reply]
>     * I'm also thinking about a little Mobile App to control the desk, maybe also adding a timer function to remind users to change position after a while. Again, I have to figure out a solution first, will share, might take a while.

Btw, the little mobile app for Android i created in the meantime. It's very nice to quickly move the desk with one tap on an icon on the mobile phone. 

Also I did an integration with my calender where an automation automatically drives the table to the right standing position when i have a web conf call.

**Reply by bjornand1** [reply]
And I use iPhone so an Android won't help either... Hope this is possible without access to either Windows or Android?

**Reply by AnAnalogGuy** [reply]
> And I use iPhone so an Android won't help either... Hope this is possible without access to either Windows or Android?

The Mobile App has nothing to do with the integration, it's only an addition to my setup. So no worries in this regard.

**Reply by AnAnalogGuy** [reply]
> Thanks for your extremely quick reply! I would love to follow the guide, but I am on a Mac and running Home Assistant on a Home Assistant Green, so I figured I would do this from inside the ESPHome Builder interface?

I can't help with the ESPHome Device Builder, as I'm not using it. 
You might try to ask Github Copilot for guidance there. 
However, adapting my guide should also work for iOS. In the end you need to connect the controller to the Mac and the Mac needs to see it. Then you need Python on the system. The rest is the same as in my guide.

**Reply by bjornand1** [reply]
Thanks, then I'll see how I can modify your guide!

**Comment by calvinbui** [comment]
I tried connecting a knock-off D1 Mini to the Megadesk today but to no avail. Can anyone notice what I've done wrong? I purchased it in February, so it should've arrived with a serial enabled firmware.

I definitely wired it up correctly with a JST wire.

```
MISO -> RX (Yellow)
SCK -> TX (Red)
GND -> GND (Black)
```

D1 Mini Rear

![image](https://user-images.githubusercontent.com/3604363/229032572-393c9a1e-0cf5-499b-a4e3-e614632b35e1.png)

D1 Mini Front

![image](https://user-images.githubusercontent.com/3604363/229032659-506a87b4-64d2-4fd0-8820-8ee1cbac0bd5.png)

Megadesk Front

![image](https://user-images.githubusercontent.com/3604363/229032755-20cc03e4-6eb6-47fc-90b7-edd17f28f9ab.png)

Megadesk Rear

![image](https://user-images.githubusercontent.com/3604363/229032697-d8e5739e-c97c-4ffe-b776-e6e465cd738d.png)

JST

![image](https://user-images.githubusercontent.com/3604363/229032467-5e0950ec-637d-4f12-93ab-bb9042afcfa0.png)

ESPHome

![image](https://user-images.githubusercontent.com/3604363/229032373-d556bed5-f8fe-446e-acd7-e0d74dfea28b.png)

**Reply by AnAnalogGuy** [reply]
- How did you solve the voltage supply? Stepdown transformer?
- Are you using the same ground on the D1/stepdown and megadesk platines?
- Does the megadesk work when triggered via buttons?
- Some of the soldering points from what i can see on the pictures could benefit from resoldering (just to make sure you don't have a connections issue there). 
- You might want to measure TX against GND and MX against GND with a Voltmeter/Multimeter while issuing a down or up command on the D1 to check if you can see a raise of voltage indicating a signal.

**Reply by calvinbui** [reply]
> How did you solve the voltage supply? Stepdown transformer?

USB-C cable directly to the D1 Mini.

> Are you using the same ground on the D1/stepdown and megadesk platines?

I'm not sure what this means.

> Does the megadesk work when triggered via buttons?

Yes, it works when pressing the buttons physically.

> You might want to measure TX against GND and MX against GND with a Voltmeter/Multimeter while issuing a down or up command on the D1 to check if you can see a raise of voltage indicating a signal.

`RX`/`TX` I get ~3.3.

`MISO` I get a ~5 & `SCK` is very low, like 0.03.

**Reply by calvinbui** [reply]
@AnAnalogGuy you replied in the wrong place haha.

The `GND` is the black cable. Are you sure it's not required?

Here's a complete picture:

![image](https://user-images.githubusercontent.com/3604363/229038248-a94e42ed-8194-4d65-a11f-bace6d3b31d4.png)

**Reply by AnAnalogGuy** [reply]
There is one GND connection not shown in your picture and that's the one that might make the problem:

![grafik](https://user-images.githubusercontent.com/70239317/229039135-96142205-95e0-453c-84e5-a500011d048d.png)

https://en.wikipedia.org/wiki/Ground_loop_(electricity)

**Reply by calvinbui** [reply]
I removed the black cable from the connector and didn't work either. I'll look at grabbing a LM2596S and trying again. They are about $3 AUD from AliExpress.

**Reply by AnAnalogGuy** [reply]
Yes, removing GND alone won't solve the issue.

Once you have the stepdown convertor, first only connect it to power supply from the Bekant (to the 24V input connectors). 
![grafik](https://user-images.githubusercontent.com/70239317/229042503-00d7c66c-c6a9-49c8-9517-0b22bec19d29.png)

Then switch the power supply on and measure the voltage outpout of the stepdown convertor on the output connectors. 
![grafik](https://user-images.githubusercontent.com/70239317/229042013-6f82823e-b1ac-47a0-9463-0ebc01abfe9b.png)
It has a variable and adjustable restistor. While measuring the output voltage, adjust the resistor so you get out 3.3V.
Once done, you can go on connecting the rest according to @fellwell5 layout. 

And of course, have the USB cable disconnected all the time to avoid a ground loop.

**Reply by AnAnalogGuy** [reply]
There is potentially a second issue with the pins you used. 
It seems you used the pins labeled TX and RX. But these are not the pins used in the code.

![grafik](https://user-images.githubusercontent.com/70239317/229232678-74db1701-9564-4419-ab76-f26d7c444ad7.png)

Unless you didn't change the code, you would have to solder SCK on megadesk to D0 on the D1 mini and accordingly MISO to D1

![grafik](https://user-images.githubusercontent.com/70239317/229233104-d1ce5144-3a5d-4dc0-b5d0-0a60ce716e6d.png)

But please, to protect the D1 mini, wait until you got the stepdown convertor to make sure you don't create a problem with GND (as these little boards can get roasted quite easily).

**Reply by calvinbui** [reply]
I changed the uart code and it worked 👍 😲 🏆 

 i'll wait for the stepdown converter before connecting it permanently

**Comment by AnAnalogGuy** [comment]
I think the ground here isn't needed as you have ground already connected trough your red-yellow-black cable.

![grafik](https://user-images.githubusercontent.com/70239317/229036233-3d055757-e1f4-4161-9577-fc1547f5b5b9.png)

Best would be you show us a complete picture of your setup with all components and connections.

**Comment by AnAnalogGuy** [comment]
Regarding GND -> if you are using different GND on your boards (and from what you are describing, you might be) you might run into issues because you might have different electrical potentials on your boards. This might be an explanation for the problem you have.

This is why @fellwell5 solved the power supply with a stepdown transformer connected to the same ground as the megadesk:

![grafik](https://user-images.githubusercontent.com/70239317/229038344-c5979d16-24aa-49c5-aa5d-f70a721ad9c6.png)

From my point of view this would be the better approach than using a seperate USB power supply. 
The stepdown convertor used here is a LM2596S.

**Comment by davidmutton** [comment]
Hi, I have been following along and set up something similar but I'm not getting any info back from my pins. I have a megadesk v6 and I haven't updated anything, does that mean my version doesn't support serial control? Am I able to update it?

**Reply by gcormier** [reply]
Correct, v6 did not have the serial enabled. You can flash your v6 PCB to the new firmware no problem.

https://github.com/gcormier/megadesk/blob/master/DIY.md

https://github.com/gcormier/megadesk/discussions/49

**Reply by davidmutton** [reply]
Thanks for getting back to me. I'm not quite at the reflashing stage of my journey with this so I just ordered a newer board instead.

**Comment by lassebm** [comment]
If you have a few minutes to spare, I would appreciate if you could take a look at my issue with strange serial responses from Megadesk: https://github.com/gcormier/megadesk/discussions/191. Thanks!
